### DISCUSSION TITLE
Is println() and readString() supported on Serial 1 on the Raspberry Pi Pico/RP2040 system

### DISCUSSION THREAD

**Discussion by kenb2054** [earlephilhower/arduino-pico#210] [Q&A] [answered]
Hi,
I am trying to port a sketch from UNO to PICO and it involves sending and receiving text on Serial port 1.
I have created the following demonstration code which compiles and uploads ok for Pico on both Arduino Mbed OS RP2040 and the community Raspberry Pi Pico/RP2040 ver 1.8.2, however it only runs ok on the Arduino Mbed OS RP2040 system.  The code seems to hang part way through when I run it on the community Raspberry Pi Pico/RP2040 ver 1.8.2

My question is whether the Serial1.println() and Serial1.readString() is supported on Raspberry Pi Pico/RP2040 ver 1.8.2, and if so, do I need to include an additional library or use some other format to make it work as it does on the Mbed OS RP2040 system.

The operation seems to freeze around the following line:     
    String receiveMessage = Serial1.readString();
There are no error messages as it compiled ok.   The last line to provide any output to the Serial Monitor is as follows and then the loop stops responding to further input:
    Serial.println(Serial1.available());        //included for diagnostic purposes only 

    // THIS FILE IS TO DEMONSTRATE TRANSMISSION OF TEXT VIA Serial and Serial1 ON PICO. 

    void setup() {
      Serial.begin(9600);     // Serial port to computer
      Serial.setTimeout(50);
      Serial1.begin(9600);    // Serial1 port - default TX pin 0, RX pin 1.    Pins are looped back for the purpose of this demo.
      Serial1.setTimeout(50);
    }

    void loop() {

      //Read from serial monitor (Serial) and send to Serial1
      if(Serial.available()){ 
        Serial.println(Serial.available());                     //included for diagnostic purposes only
        String sendMessage = Serial.readString();   //Read string from serial monitor
        Serial1.println(sendMessage);               // Send to Serial1 TX pin which is looped back to Serial1 RX pin
        Serial.print("Message Sent to Serial1 is:  ");
        Serial.println(sendMessage);  
        delay(1000);
      }    

      //Read from Serial1 and send to Serial Monitor (Serial)
      if(Serial1.available()){                            // Read from Serial1 RX pin which is looped back from Serial1 TX pin
        Serial.println(Serial1.available());        //included for diagnostic purposes only
        String receiveMessage = Serial1.readString();          //THIS IS WHERE I SUSPECT THE PROGRAM FREEZES
        Serial.print("Message received on Serial1 is:  ");
        Serial.println(receiveMessage);        // Send to serial monitor  
      }
     }

**Comment by earlephilhower** [comment]
Yes, they should all work.  

It;s most likely a bug in the core.  I haven't done extensive input testing on the UART Serial ports, but looks like I need to.  Thanks for the example, I'll give it a go later today.  I should also probably add a loopback example like `TalkingToMyself.ino` as a sanity check I can run anytime.

Also, just FYI, in this core the `Serial` port is the UART-based one, and `Serial1` is UART0 and `Serial2` is UART1.  I'm not certain the MBED core uses the same organization.  I chose this way because 99% of Arduino users will probably be using the USB connection, not a USB<->Serial adapter.

**Reply by earlephilhower** [reply]
Ahhh, I think I see the problem.  `Serial.setTimeout()` is not checked in either USB or UART read!  I did not try that before, and it's required to terminate your `Serial1.readString()` call, OTW it will hang forever as you found out.

**Reply by earlephilhower** [reply]
@kenb2054 Please try the 1.8.3 release, just uploaded.  I ran your  test and it came out fine:
````
13:59:08.892 -> Message Sent to Serial1 is:  techno
13:59:08.892 -> 
13:59:09.886 -> 1
13:59:09.919 -> Message received on Serial1 is:  techno
13:59:09.919 -> 
13:59:09.919 -> 
13:59:13.331 -> 8
13:59:13.398 -> Message Sent to Serial1 is:  gamama
13:59:13.398 -> 
13:59:14.392 -> 1
13:59:14.425 -> Message received on Serial1 is:  gamama
13:59:14.425 -> 
````

**Comment by kenb2054** [comment, marked as answer]
@earlephilhower.   Thanks very much for your quick response.   That has solved the problem.

**Comment by smileshop** [comment]
@earlephilhower.  Have you ever try receiving string longer than 32 bytes through Serial1(UART0) or Serial2 (UART1)?  I have tried 125 characters. Just receive some of them with readString( ) or readStringUntil( ) or add in each character into an array.

**Reply by smileshop** [reply]
Here is a sample string test uart receiving.

# 07-25-1901 11:21:56 670 25.0
& 0.0 0.0 3.910E-02
& 50.0 180.0 2.279E-02
& 25.0 0.0 7.000E-03
*** TRANSFER COMPLETE

**Reply by smileshop** [reply]
I fixed the problem. 
1. Minimize delay in the Arduino loop.
2. In stead of read all incoming strings but read each line by detecting cr and lf characters. Add each line into the array of characters.
